### PR TITLE
Add issue comment when moving issues from one column to another of the project

### LINKED ISSUE #27278: Add the column movement information on the timeline of the Issue detail page.
### Feature Description

We rely on Gitea’s Projects feature for actual project management. However, we are currently unable to see the changes in the project column from the timeline on the issue details page. In other words, we cannot see when and by whom an issue was moved from one column to another. This is causing confusion in our project management.

I think if this feature is added, “Projects” will become more useful and benefit more Gitea users.

### Screenshots

_No response_

### PR DESCRIPTION
**Pull request by lunny**
Fix #27278
Replace #27816

This PR adds a meta-comment for an issue when dragging an issue from one column to another of a project.

image

### CODE REVIEW

File: templates/repo/issue/view_content/comments.tmpl [hunk 1]
```diff
@@ -600,6 +600,22 @@
...
 			</div>
 			{{end}}
+		{{else if eq .Type 31}}
+			{{if not $.UnitProjectsGlobalDisabled}}
+			<div class="timeline-item event" id="{{.HashTag}}">
+				<span class="badge">{{svg "octicon-project"}}</span>
+				{{template "shared/user/avatarlink" dict "user" .Poster}}
+				<span class="text grey muted-links">
+					{{template "shared/user/authorlink" .Poster}}
+					{{$newProjectDisplayHtml := "Unknown Project"}}
```
Comment (silverwind) [inline templates/repo/issue/view_content/comments.tmpl]: Translate string?

File: models/issues/comment.go [hunk 2]
```diff
@@ -974,34 +993,37 @@ type CreateCommentOptions struct {
...
+	ReviewID           int64
+	Content            string
+	Attachments        []string // UUIDs of attachments
+	RefRepoID          int64
+	RefIssueID         int64
+	RefCommentID       int64
+	RefAction          references.XRefAction
+	RefIsPull          bool
+	IsForcePush        bool
+	Invalidated        bool
```
Comment (silverwind) [inline models/issues/comment.go:1034]: And this is why `go fmt` sucks. Can't even tell what was changed.

Comment (yardenshoham) [reply models/issues/comment.go:1034]: image
Using https://mergease.com/welcome

Comment (silverwind) [reply models/issues/comment.go:1034]: Hide whitespace option on GH also does it, but the bigger problem is also that `gofmt` also destroys valuable line attribution and therefor blame info, it just sucks all around 😆

File: models/migrations/v1_22/v287.go [hunk 3]
```diff
@@ -0,0 +1,22 @@
+// Copyright 2023 The Gitea Authors. All rights reserved.
```
Comment (yp05327) [inline models/migrations/v1_22/v287.go]: 2024?

File: services/projects/issue.go [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+			return err
+		}
+
+		project, err := project_model.GetProjectByID(ctx, column.ProjectID)
+		if err != nil {
+			return err
+		}
+
+		for sorting, issueID := range sortedIssueIDs {
+			_, err = sess.Exec("UPDATE `project_issue` SET project_board_id=?, sorting=? WHERE issue_id=?", column.ID, sorting, issueID)
```
Comment (yp05327) [inline services/projects/issue.go]: Old codes use SQL directly. We can use builder now.

Comment (lunny) [reply services/projects/issue.go]: Both are OK for me.

File: templates/repo/issue/view_content/comments.tmpl (hunk 1 above)
Comment (lunny) [reply templates/repo/issue/view_content/comments.tmpl]: changed.

**Review by silverwind — APPROVED** [review]

File: templates/repo/issue/view_content/comments.tmpl [hunk 5]
```diff
@@ -601,6 +601,22 @@
...
 			</div>
 			{{end}}
+		{{else if eq .Type 31}}
+			{{if not $.UnitProjectsGlobalDisabled}}
+			<div class="timeline-item event" id="{{.HashTag}}">
+				<span class="badge">{{svg "octicon-project"}}</span>
+				{{template "shared/user/avatarlink" dict "user" .Poster}}
+				<span class="text grey muted-links">
+					{{template "shared/user/authorlink" .Poster}}
+					{{$newProjectDisplayHtml := .CommentMetaData.ProjectTitle|Safe}}
```
Comment (wxiaoguang) [inline templates/repo/issue/view_content/comments.tmpl]: XSS.

In most case, do not use "Escape" or "Safe" in templates.

Comment (lunny) [reply templates/repo/issue/view_content/comments.tmpl]: I think I just copied them from Type 30.

Comment (wxiaoguang) [reply templates/repo/issue/view_content/comments.tmpl]: Type 30 never does `{{$newProjectDisplayHtml := .CommentMetaData.ProjectTitle|Safe}}`

Comment (lunny) [reply templates/repo/issue/view_content/comments.tmpl]: 1dfb66c706eac2d54c49203b49deb95f3f5265cc

**Comment by wxiaoguang** [comment]
Give me some more time. I will completely remove these annoying "Escape" and "Safe" from the templates.

**Comment by wxiaoguang** [comment]
->  Refactor Safe modifier #29392

File: templates/repo/issue/view_content/comments.tmpl [hunk 6]
```diff
@@ -612,6 +612,22 @@
...
 			</div>
 			{{end}}
+		{{else if eq .Type 31}}
+			{{if not $.UnitProjectsGlobalDisabled}}
+			<div class="timeline-item event" id="{{.HashTag}}">
+				<span class="badge">{{svg "octicon-project"}}</span>
+				{{template "shared/user/avatarlink" dict "user" .Poster}}
+				<span class="text grey muted-links">
+					{{template "shared/user/authorlink" .Poster}}
+					{{$newProjectDisplayHtml := .CommentMetaData.ProjectTitle | Escape}}
```
Comment (wxiaoguang) [inline templates/repo/issue/view_content/comments.tmpl]:
```suggestion
					{{$newProjectDisplayHtml := .CommentMetaData.ProjectTitle}}
```

File: templates/repo/issue/view_content/comments.tmpl [hunk 7]
```diff
@@ -612,6 +612,22 @@
...
+					{{if .Project}}
+						{{$trKey := printf "projects.type-%d.display_name" .Project.Type}}
+						{{$newProjectDisplayHtml = HTMLFormat `%s <a href="%s"><span data-tooltip-content="%s">%s</span></a>` (svg .Project.IconName) (.Project.Link ctx) (ctx.Locale.Tr $trKey) .Project.Title}}
+					{{end}}
+					{{ctx.Locale.Tr "repo.issues.move_to_column_of_project" (.CommentMetaData.ProjectColumnTitle|Escape) $newProjectDisplayHtml $createdStr}}
```
Comment (wxiaoguang) [inline templates/repo/issue/view_content/comments.tmpl]:
```suggestion
					{{ctx.Locale.Tr "repo.issues.move_to_column_of_project" .CommentMetaData.ProjectColumnTitle $newProjectDisplayHtml $createdStr}}
```

File: templates/repo/issue/view_content/comments.tmpl [hunk 8]
```diff
@@ -612,6 +612,22 @@
...
+				<span class="badge">{{svg "octicon-project"}}</span>
+				{{template "shared/user/avatarlink" dict "user" .Poster}}
+				<span class="text grey muted-links">
+					{{template "shared/user/authorlink" .Poster}}
+					{{$newProjectDisplayHtml := .CommentMetaData.ProjectTitle}}
+					{{if .Project}}
+						{{$trKey := printf "projects.type-%d.display_name" .Project.Type}}
+						{{$newProjectDisplayHtml = HTMLFormat `%s <a href="%s"><span data-tooltip-content="%s">%s</span></a>` (svg .Project.IconName) (.Project.Link ctx) (ctx.Locale.Tr $trKey) .Project.Title}}
+					{{end}}
+					{{ctx.Locale.Tr "repo.issues.move_to_column_of_project" .CommentMetaData.ProjectColumnTitle $newProjectDisplayHtml $createdStr}}
```
Comment (wxiaoguang) [inline templates/repo/issue/view_content/comments.tmpl]:
```suggestion
					{{$newProjectDisplay := .CommentMetaData.ProjectTitle}}
					{{if .Project}}
						{{$trKey := printf "projects.type-%d.display_name" .Project.Type}}
						{{$newProjectDisplay = HTMLFormat `%s <a href="%s"><span data-tooltip-content="%s">%s</span></a>` (svg .Project.IconName) (.Project.Link ctx) (ctx.Locale.Tr $trKey) .Project.Title}}
					{{end}}
					{{ctx.Locale.Tr "repo.issues.move_to_column_of_project" .CommentMetaData.ProjectColumnTitle $newProjectDisplay $createdStr}}
```

File: templates/repo/issue/view_content/comments.tmpl [hunk 9]
```diff
@@ -597,6 +597,22 @@
...
+			{{if not $.UnitProjectsGlobalDisabled}}
+			<div class="timeline-item event" id="{{.HashTag}}">
+				<span class="badge">{{svg "octicon-project"}}</span>
+				{{template "shared/user/avatarlink" dict "user" .Poster}}
+				<span class="text grey muted-links">
+					{{template "shared/user/authorlink" .Poster}}
+					{{$newProjectDisplay := .CommentMetaData.ProjectTitle}}
+					{{if .Project}}
+						{{$trKey := printf "projects.type-%d.display_name" .Project.Type}}
+						{{$newProjectDisplay = HTMLFormat `%s <a href="%s"><span data-tooltip-content="%s">%s</span></a>` (svg .Project.IconName) (.Project.Link ctx) (ctx.Locale.Tr $trKey) .Project.Title}}
```
Comment (silverwind) [inline templates/repo/issue/view_content/comments.tmpl:617]: Side note: Interesting helper `HTMLFormat`, I wasn't aware of it yet.

File: models/migrations/v1_23/v294.go [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+import (
+	"xorm.io/xorm"
+)
+
+// CommentMetaData stores metadata for a comment, these data will not be changed once inserted into database
+type CommentMetaData struct {
+	ProjectColumnID   int64  `json:"project_column_id"`
+	ProjectColumnName string `json:"project_column_name"`
+	ProjectName       string `json:"project_name"`
+}
```
Comment (yp05327) [inline models/migrations/v1_23/v294.go]: It seems that the field name is updated above but not be changed here?
https://github.com/go-gitea/gitea/pull/29311/commits/0be30d9e3a2092f4873ae59bafd1907296b50f5d#diff-2a26684f2135ebd22622ba68351130d7b7ae824ac88728cbbded62123a718d15R224-R225

Comment (lunny) [reply models/migrations/v1_23/v294.go]: Fixed.

Comment (yp05327) [reply models/migrations/v1_23/v294.go]: You fixed `ProjectName`
But how about `ProjectColumnName` vs `ProjectColumnTitle` ?

Comment (yp05327) [reply models/migrations/v1_23/v294.go]: image
image

File: models/migrations/v1_23/v297.go [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+// CommentMetaData stores metadata for a comment, these data will not be changed once inserted into database
+type CommentMetaData struct {
+	ProjectColumnID    int64  `json:"project_column_id"`
+	ProjectColumnTitle string `json:"project_column_title"`
+	ProjectTitle       string `json:"project_title"`
+}
+
+func AddCommentMetaDataColumn(x *xorm.Engine) error {
+	type Comment struct {
+		CommentMetaData CommentMetaData `xorm:"JSON TEXT"` // put all non-index metadata in a single field
```
Comment (yp05327) [inline models/migrations/v1_23/v297.go]:
```suggestion
		CommentMetaData *CommentMetaData `xorm:"JSON TEXT"` // put all non-index metadata in a single field
```

File: models/migrations/v1_22/v287.go [hunk 12]
```diff
@@ -1,6 +1,5 @@
 // Copyright 2023 The Gitea Authors. All rights reserved.
 // SPDX-License-Identifier: MIT
-
```
Comment (yp05327) [inline models/migrations/v1_22/v287.go]: ???

File: services/projects/issue.go [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+		}
+
+		for sorting, issueID := range sortedIssueIDs {
+			_, err = db.Exec(ctx, "UPDATE `project_issue` SET project_board_id=?, sorting=? WHERE issue_id=?", column.ID, sorting, issueID)
+			if err != nil {
+				return err
+			}
+
+			var curIssue *issues_model.Issue
+			for _, issue := range issues {
```
Comment (yp05327) [inline services/projects/issue.go]: Can `issues` be a map? Then there's no need to do for loop every time here.

File: routers/web/org/projects.go [hunk 14]
```diff
@@ -600,7 +602,7 @@ func MoveIssues(ctx *context.Context) {
 		}
 	}
 
-	if err = project_model.MoveIssuesOnProjectColumn(ctx, column, sortedIssueIDs); err != nil {
```
Comment (yp05327) [inline routers/web/org/projects.go:604]: It seems that `project_model.MoveIssuesOnProjectColumn` isn't removed.

Comment (lunny) [reply routers/web/org/projects.go:604]: f1b942e

File: services/projects/issue.go (hunk 13 above)
Comment (lunny) [reply services/projects/issue.go]: f1b942e

File: models/migrations/v1_22/v287.go (hunk 12 above)
Comment (lunny) [reply models/migrations/v1_22/v287.go]: f1b942e

File: services/projects/issue.go [hunk 15]
```diff
@@ -0,0 +1,76 @@
...
+			issuesMap[issue.ID] = issue
+		}
+
+		for sorting, issueID := range sortedIssueIDs {
+			_, err = db.Exec(ctx, "UPDATE `project_issue` SET project_board_id=?, sorting=? WHERE issue_id=?", column.ID, sorting, issueID)
+			if err != nil {
+				return err
+			}
+
+			curIssue := issuesMap[issueID]
```
Comment (yp05327) [inline services/projects/issue.go]: It would cause panic, if we don't check whether it is exist?

Comment (lunny) [reply services/projects/issue.go]: e1d45e15be24e8f087e38e91d17f45dcb40a7868

**Review by yp05327 — APPROVED** [review]

**Pull request merged**